pkg/monitor/monitor.go

```diff
@@ -0,0 +1,101 @@
+package monitor
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"log/slog"
+	"time"
+
+	"github.com/redhat-appstudio/rhtap-cli/pkg/k8s"
+
+	"k8s.io/cli-runtime/pkg/resource"
+)
+
+// monitorQueueFn is a function type for monitoring a specific resource.
+type monitorQueueFn func() error
+
+// Monitor is the monitoring actor which collects interesting resources from a
+// Helm Chart release payload, and monitors them until they are ready. The
+// monitoring is executed with a queue of functions, which are executed in order
+// until the queue is empty or timeout is reached.
+type Monitor struct {
+	logger *slog.Logger  // application logger
+	kube   k8s.Interface // kubernetes client
+
+	queue []monitorQueueFn // monitor function queue
+}
+
+var _ Interface = &Monitor{}
+
+// Collect inspects the resource and adds a monitoring function to the queue.
+func (m *Monitor) Collect(ctx context.Context, r *resource.Info) error {
+	if r.Object == nil {
+		return fmt.Errorf("resource object is nil")
+	}
+
+	gvk := r.Object.GetObjectKind().GroupVersionKind()
+	gv := gvk.GroupVersion().String()
+
+	logger := m.logger.With(
+		"gv", gv,
+		"kind", gvk.Kind,
+		"name", r.Name,
+		"namespace", r.Namespace,
+	)
+	logger.Debug("Inspecting resource for monitoring...")
+
+	switch fmt.Sprintf("%s/%s", gv, gvk.Kind) {
+	case "project.openshift.io/v1/ProjectRequest":
+		logger.Debug("ProjectRequest detected, waiting for namespace creation...")
+		fn, err := AssertNamespaceFn(ctx, m.logger, m.kube, r.Name)
+		if err != nil {
+			return err
+		}
+		m.queue = append(m.queue, fn)
+	}
+	return nil
+}
+
+// Watch waits for all monitoring functions to complete, or until the timeout is
+// reached. Returns error if the queue is not empty after timeout.
+func (m *Monitor) Watch(timeout time.Duration) error {
+	start := time.Now()
+	logger := m.logger.With("timeout", timeout.String(), "start", start.String())
+	for {
+		for i := 0; i < len(m.queue); {
```

Comment:
I don't understand the value of this nested loop.

Reply:
What I mean is that the following code seems to lead to the same result while being easier to read.

```
	for len(m.queue) > 0 {
		if time.Since(start) >= timeout {
			return errors.New("timeout reached")
		}

		if err := m.queue[0](); err == nil {
			logger.Debug("Monitor function succeeded!")
			m.queue = m.queue[1:]
		} else {
			logger.Debug("Monitor function failed!")
			time.Sleep(300 * time.Millisecond)	
		}
	}
	logger.Debug("Monitoring complete, queue is empty!")
	return nil
```

Reply:
I would also bump the delay to 2s.

Reply:
Okay, sure! Good suggestions here, please consider latest changes.